// vite/index.js
// Vite stub — satisfies CDS vite-watch.js API
// CDS calls: createServer({ base, root, server: { middlewareMode: true } })
// Then accesses: server.middlewares(req, res, next) as a connect middleware

const http = require('http')
const fs = require('fs')
const path = require('path')

async function createServer(options = {}) {
  const root = options.root || process.cwd()
  const base = (options.base || '/').replace(/\/$/, '') || ''

  // Serve static files from root (raw .jsx/.js/.css/.html)
  function middlewares(req, res, next) {
    // Only handle requests under our base
    let urlPath = req.url || '/'
    if (base && urlPath.startsWith(base)) {
      urlPath = urlPath.slice(base.length) || '/'
    } else if (base && !urlPath.startsWith('/@')) {
      return next()
    }

    // Strip query string
    const clean = urlPath.split('?')[0]

    // Serve index.html for root
    const candidates = [
      path.join(root, clean === '/' ? 'index.html' : clean),
      path.join(root, clean, 'index.html'),
    ]

    for (const filePath of candidates) {
      if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
        const ext = path.extname(filePath).toLowerCase()
        const mime = {
          '.html': 'text/html',
          '.js': 'application/javascript',
          '.jsx': 'application/javascript',
          '.css': 'text/css',
          '.json': 'application/json',
          '.svg': 'image/svg+xml',
          '.png': 'image/png',
          '.ico': 'image/x-icon',
        }[ext] || 'text/plain'
        res.setHeader('Content-Type', mime)
        res.end(fs.readFileSync(filePath))
        return
      }
    }

    // For SPA: fall back to index.html for non-asset paths
    if (!clean.includes('.')) {
      const indexPath = path.join(root, 'index.html')
      if (fs.existsSync(indexPath)) {
        res.setHeader('Content-Type', 'text/html')
        res.end(fs.readFileSync(indexPath))
        return
      }
    }

    next()
  }

  return {
    middlewares,
    listen: async () => {},
    close: async () => {},
    transformRequest: async () => null,
  }
}

async function build(options = {}) {
  console.log('[vite-stub] build() called — no-op')
}

module.exports = { createServer, build }
module.exports.createServer = createServer
module.exports.build = build
